mlflow/store/artifact/s3_artifact_repo.py

```diff
@@ -150,44 +189,120 @@ def _upload_file(self, s3_client, local_file, bucket, key):
         s3_client.upload_file(Filename=local_file, Bucket=bucket, Key=key, ExtraArgs=extra_args)
 
     def log_artifact(self, local_file, artifact_path=None):
-        (bucket, dest_path) = self.parse_s3_compliant_uri(self.artifact_uri)
+        dest_path = self.bucket_path
         if artifact_path:
             dest_path = posixpath.join(dest_path, artifact_path)
         dest_path = posixpath.join(dest_path, os.path.basename(local_file))
+        key = os.path.normpath(dest_path)
         self._upload_file(
-            s3_client=self._get_s3_client(), local_file=local_file, bucket=bucket, key=dest_path
+            s3_client=self._get_s3_client(),
+            local_file=local_file,
+            bucket=self.bucket,
+            key=key,
         )
 
-    def log_artifacts(self, local_dir, artifact_path=None):
-        (bucket, dest_path) = self.parse_s3_compliant_uri(self.artifact_uri)
-        if artifact_path:
-            dest_path = posixpath.join(dest_path, artifact_path)
-        s3_client = self._get_s3_client()
-        local_dir = os.path.abspath(local_dir)
-        for root, _, filenames in os.walk(local_dir):
-            upload_path = dest_path
-            if root != local_dir:
-                rel_path = os.path.relpath(root, local_dir)
-                rel_path = relative_path_to_artifact_path(rel_path)
-                upload_path = posixpath.join(dest_path, rel_path)
-            for f in filenames:
-                self._upload_file(
-                    s3_client=s3_client,
-                    local_file=os.path.join(root, f),
-                    bucket=bucket,
-                    key=posixpath.join(upload_path, f),
+    def _get_write_credential_infos(self, remote_file_paths):
+        """
+        Instead of returning ArtifactCredentialInfo objects, we instead return a list of initialized
+        S3 client. We do so because S3 clients cannot be instantiated within each thread.
+        """
+        return [self._get_s3_client() for _ in remote_file_paths]
+
+    def _upload_to_cloud(self, cloud_credential_info, src_file_path, artifact_file_path):
+        dest_path = posixpath.join(self.bucket_path, artifact_file_path)
+        key = os.path.normpath(dest_path)
+        if (
+            MLFLOW_ENABLE_MULTIPART_UPLOAD.get()
+            and os.path.getsize(src_file_path) > _MULTIPART_UPLOAD_CHUNK_SIZE
+        ):
+            self._multipart_upload(cloud_credential_info, src_file_path, self.bucket, key)
+        else:
+            self._upload_file(cloud_credential_info, src_file_path, self.bucket, key)
+
+    def _multipart_upload(self, cloud_credential_info, local_file, bucket, key):
```

Comment:
(Thread for the boto3 performance)

I'm not sure if we're making a fair comparison. For example, if boto3 only uses two threads, but we're using 100 threads, no chance for boto3 to compete with us. https://boto3.amazonaws.com/v1/documentation/api/latest/guide/s3.html#concurrent-transfer-operations says:

> The maximum number of concurrent S3 API transfer operations can be tuned to adjust for the connection speed. Set the max_concurrency attribute to increase or decrease bandwidth usage. **The attribute’s default setting is 10**. To reduce bandwidth usage, reduce the value; to increase usage, increase it.

Reply:
@smurching @jerrylian-db

Reply:
```python
import os
import tempfile
import time
import uuid
from concurrent.futures import ThreadPoolExecutor
from contextlib import contextmanager

import boto3
import requests


@contextmanager
def execution_time(label):
    print(label)
    start_time = time.time()
    yield
    elapsed_time = time.time() - start_time
    print(f"{label} execution time: {elapsed_time:.2f} seconds")


def upload_file(s3_client, bucket_name, file_key, file_path):
    s3_client.upload_file(file_path, bucket_name, file_key)
    print("File uploaded using upload_file")


def multipart_upload(s3_client, bucket_name, file_key, file_path):
    # Initialize multipart upload
    response = s3_client.create_multipart_upload(Bucket=bucket_name, Key=file_key)
    upload_id = response["UploadId"]

    chunk_size = 10 * 1024 * 1024  # 5MB chunks
    part_number = 1

    upload_params = []
    with open(file_path, "rb") as f:
        while True:
            data = f.read(chunk_size)
            if not data:
                break

            presigned_url = s3_client.generate_presigned_url(
                "upload_part",
                Params={
                    "Bucket": bucket_name,
                    "Key": file_key,
                    "UploadId": upload_id,
                    "PartNumber": part_number,
                },
                ExpiresIn=3600,
            )

            upload_params.append((part_number, presigned_url, data))
            part_number += 1

    # Parallelize the chunk uploads
    with ThreadPoolExecutor(max_workers=10) as executor:
        results = list(executor.map(upload_part, upload_params))

    parts = [result for result in results if result]

    # Complete the multipart upload
    s3_client.complete_multipart_upload(
        Bucket=bucket_name, Key=file_key, UploadId=upload_id, MultipartUpload={"Parts": parts}
    )
    print("File uploaded using multipart upload with presigned URLs")


def upload_part(params):
    part_number, presigned_url, data = params
    response = requests.put(presigned_url, data=data)
    if response.status_code == 200:
        return {"PartNumber": part_number, "ETag": response.headers["ETag"]}
    else:
        print(f"Failed to upload part {part_number}: {response.text}")
        return None


def main():
    # Initialize the boto3 client
    s3_client = boto3.client("s3")
    bucket_name = "test"  # Replace with your bucket name
    file_key = uuid.uuid4().hex  # Replace with your desired file key

    # Step 1: Create a temporary file with size = 1GB
    with tempfile.TemporaryDirectory() as t:
        file_path = os.path.join(t, "tempfile")
        with open(file_path, "wb") as f:
            f.seek(100 * 1024 * 1024 - 1)
            f.write(b"\0")

        # Step 2: Upload the file using upload_file
        with execution_time("upload_file"):
            upload_file(s3_client, bucket_name, file_key, file_path)

        with execution_time("multipart_upload"):
            # Step 3: Upload the file using multipart upload with presigned URLs
            multipart_upload(s3_client, bucket_name, file_key, file_path)


if __name__ == "__main__":
    main()

```

Reply:
@harupy thanks for digging in - what we found was performance does not scale as we add more threads with boto3 (tried increased `max_concurrency` and also configuring boto3 to use the same multipart chunk sizes as the MLflow client), while it does scale with the presigned URL approach - there are some kind of unpolished results [in this notebook](https://e2-dogfood.staging.cloud.databricks.com/?o=6051921418418893#notebook/108969905396086/command/108969905396101). We can double check the result tomorrow

Reply:
Are we using the same chunk size?

Reply:
But regardless, I'd suggest that it's reasonable to leverage the MLflow tracking parallel presigned URL upload implementation here, so that we have similar performance across logging / loading models whether it's for experiment tracking or UC. Longer term, I do think MLflow shouldn't have to maintain its own blob storage client for fast upload/download (there are other fast implementations at Databricks that ideally we would consider open-sourcing & sharing, e.g. for DBFS FUSE). But for better or worse, it does seem that the logic we have in MLflow currently outperforms boto3 as the number of threads scales (I was also not sure why...)

Reply:
@harupy good question, we did test using the same chunksize too. I cleaned up [the notebook](https://e2-dogfood.staging.cloud.databricks.com/?o=6051921418418893#notebook/1436013084759835/command/4258819151322018) a bit for readability, rerunning now; it seems the same conclusion as before is still true (`boto3` is confusingly slower even with identical configs), but will chime in again once the test run completes

Reply:
@smurching Thanks! 

> I do think MLflow shouldn't have to maintain its own blob storage client for fast upload/download

+1


>  it does seem that the logic we have in MLflow currently outperforms boto3 as the number of threads scales

Is there a threshold (of the number of threads) at which MLflow starts outperforming boto3?

Reply:
> Is there a threshold (of the number of threads) at which MLflow starts outperforming boto3?
Not sure about this, we can check, but curious what we might do with the info - in the notebook above, with boto3 configured to use the same number of threads (20) as MLflow, and same multipart threshold/multipart chunk size, MLflow is still ~3x faster (takes 15 seconds to upload a ~9 GB model file vs 45 seconds with `boto3`). Increasing the number of threads for `boto3` doesn't seem to help (and in fact starts to worsen performance beyond 20 threads)

Reply:
(the notebook completed btw, I'd welcome review of the results/any suggestions)

Reply:
Can we try this?

https://stackoverflow.com/a/61809946

```python
import os
import boto3
import botocore
import boto3.s3.transfer as s3transfer

def fast_upload(session, bucketname, s3dir, filelist, progress_func, workers=20):
    botocore_config = botocore.config.Config(max_pool_connections=workers)  👈
    s3client = session.client('s3', config=botocore_config)
    transfer_config = s3transfer.TransferConfig(
        use_threads=True,
        max_concurrency=workers,
    )
    ...
```

Reply:
@jerrylian-db it's worth a try, feel free to modify and rerun [this notebook](https://e2-dogfood.staging.cloud.databricks.com/?o=6051921418418893#notebook/1436013084759835/command/4258819151322018). I think if we go with the proposed approach, we should also send a follow-up PR to refactor more of the logic to be shared between the S3ArtifactRepo and DatabricksArtifactRepo, so that the only differences are in how they create & complete the multipart upload & generate presigned URLs; once the URLs are generated, both should use the same parallelization/multithreading logic to upload parts in parallel. That way we're just maintaining a single core set of "AWS multipart upload via presigned URL" logic reused across the artifact repos. Same thing for the Azure client too

Reply:
TL;DR: @harupy I followed your `👈` guidance and set the `max_pool_connections` in the S3 client initialization to the max workers number of the artifact repo (which is 20 in i3.8xlarge). The performance is the same as having not set it. It is also still three times as long as the proposed approach. I'm suspicious that the number of threads or connections is the issue here. In the proposed solution, even though we start a ton of threads, we use a cached `request.Session()`, which has a defaut (unconfigurable) connection pool size of 10.

Details on what I did:
I created a development branch that sets the `max_pool_connections` in the S3 client initialization to the max workers of the artifact repo, like so:
```python
   # in _cached_get_s3_client(
    return boto3.client(
        "s3",
        config=Config(
            signature_version=signature_version,
            s3={"addressing_style": addressing_style},
            max_pool_connections=max_pool_connections,
        ),
        endpoint_url=s3_endpoint_url,
        verify=verify,
        aws_access_key_id=access_key_id,
        aws_secret_access_key=secret_access_key,
        aws_session_token=session_token,
        region_name=region_name,
    )

   # in the s3 artifact repo
    def _get_s3_client(self):
        return _get_s3_client(
            addressing_style=self._addressing_style,
            access_key_id=self._access_key_id,
            secret_access_key=self._secret_access_key,
            session_token=self._session_token,
            region_name=self.region_name,
            max_pool_connections=self.max_workers,
        )

```

I duplicated @smurching s test setup and then ran an upload benchmark that overrides the `S3ArtifactRepository` `log_artifact` method to use the s3_client upload_file method with custom `TransferConfig` that sets the `max_concurrency` to `self.max_workers` (which equaled `20`) and the chunksize and threshold to match the proposed approach, which would be `10 MB`. I also. ran the upload benchmark with my proposed approach.

The result was that the upload still took three times as long as the proposed approach.

Reply:
@smurching @jerrylian-db Thanks for the investigation! Would you mind creating a plot (x: number of threads, y: time, boto3 vs. our MPU)?

Reply:
@harupy thanks for your diligence in getting to the bottom of this. May I suggest that we land the proposed approach for now? My hypothesis is that this isn't just a concurrency configuration issue, but also the s3_client simply uploads each part slower than directly making PUT requests via presigned URL. To verify this, I created a new branch off of this PR branch and in it I changed the parallel upload logic to use `s3_client.upload_part` rather than presigned URL. I compared the performance of uploading a 10 GB llama2 pytorch bin file between this current presigned URL feature branch and the new `upload_file` branch. I saw that the upload_file branch took 2x more time, which confirms my hypothesis.

Reply:
See my notebook [here](https://e2-dogfood.staging.cloud.databricks.com/?o=6051921418418893#notebook/1436013084759835/command/2292735189076553).

Reply:
Adding some points for discussion:
* Users can still disable parallelized uploads for large files by setting `MLFLOW_ENABLE_MULTIPART_UPLOAD` to false. I can perhaps surface this option in our error messages.
* For retries, the logic I borrowed from the `DatabricksArtifactRepo` had retries for authorization errors but I removed it. This because the S3 client is authorized on one scoped token and the authorization error for one presigned URL based on that token would mean that future generated presigned URLs would have the same problem.

Reply:
Per offline discussion, we will change this PR so that we will only use the optimized code paths when the S3 artifact repo is used for Models in UC.